crates/sui-core/src/authority.rs

```diff
@@ -476,16 +479,24 @@ impl AuthorityState {
             .map(|(_, obj)| obj.compute_object_reference())
             .sorted()
             .collect();
+
+        let shared_objects_ids = shared_object_refs.iter().map(|(id, _, _)| *id);
+        let tombstones = self.database.tombstones_exist(shared_objects_ids)?;
+
         if !shared_object_refs.is_empty() && !certificate.data.kind.is_system_tx() {
             // If the transaction contains shared objects, we need to ensure they have been scheduled
             // for processing by the consensus protocol.
             // There is no need to go through consensus for system transactions that can
             // only be executed at a time when consensus is turned off.
             // TODO: Add some assert here to make sure consensus is indeed off with is_system_tx.
-            self.check_shared_locks(&transaction_digest, &shared_object_refs)
+            self.check_shared_locks(&transaction_digest, &shared_object_refs, &tombstones)
                 .await?;
         }
 
+        // If there is a tombstone for even a single shared object, we need unlock all
+        // owned-objects and increment the sequence number of all existing shared-objects.
+        // TODO
```

Comment:
If a shared object has a tombstone (i.e., its ID is in the set `tombstones` line 484):
* increase the version number of each shared object in the certificate
* unlock every owned-object and increment their sequence number
* charge gas
* tell the user that the transaction failed because one or more shared object is dead